persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,680 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections #-}
+
+module Database.Persist.Quasi.Internal.ModelParser
+    ( SourceLoc (..)
+    , Token (..)
+    , tokenContent
+    , anyToken
+    , ParsedEntityDef
+    , parsedEntityDefComments
+    , parsedEntityDefEntityName
+    , parsedEntityDefIsSum
+    , parsedEntityDefEntityAttributes
+    , parsedEntityDefFieldAttributes
+    , parsedEntityDefExtras
+    , parsedEntityDefSpan
+    , parseSource
+    , memberBlockAttrs
+    , ParseResult
+    , CumulativeParseResult (..)
+    , toCumulativeParseResult
+    , renderErrors
+    , runConfiguredParser
+    ) where
+
+import Control.Monad.Trans.State
+import Data.List (intercalate)
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NEL
+import qualified Data.Map as M
+import Data.Maybe
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Void
+import Database.Persist.Types
+import Database.Persist.Types.SourceSpan
+import Language.Haskell.TH.Syntax (Lift)
+import Replace.Megaparsec (sepCap)
+import Text.Megaparsec hiding (Token)
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+-- We'll augment the parser with extra state to accumulate comments seen during parsing.
+-- Comments are lexed as whitespace, but will be used to generate documentation later.
+type ExtraState = [(SourcePos, CommentToken)]
+
+type Parser a =
+    StateT
+        ExtraState
+        (ParsecT Void String (Either (ParseErrorBundle String Void)))
+        a
+type InternalParseResult a =
+    Either (ParseErrorBundle String Void) (a, ExtraState)
+type ParseResult a = Either (ParseErrorBundle String Void) a
+
+-- | Run a parser using a provided ExtraState
+-- @since 2.16.0.0
+runConfiguredParser
+    :: ExtraState
+    -> Parser a
+    -> String
+    -> String
+    -> InternalParseResult a
+runConfiguredParser acc parser fp s = do
+    (_internalState, parseResult) <-
+        runParserT' (runStateT parser acc) initialInternalState
+    parseResult
+  where
+    initialSourcePos =
+        SourcePos
+            { sourceName = fp
+            , sourceLine = pos1
+            , sourceColumn = pos1
+            }
+    initialPosState =
+        PosState
+            { pstateInput = s
+            , pstateOffset = 0
+            , pstateSourcePos = initialSourcePos
+            , -- for legacy compatibility, we treat each tab as a single unit of whitespace
+              pstateTabWidth = pos1
+            , pstateLinePrefix = ""
+            }
+    initialInternalState =
+        State
+            { stateInput = s
+            , stateOffset = 0
+            , statePosState = initialPosState
+            , stateParseErrors = []
+            }
+
+-- @since 2.16.0.0
+data CumulativeParseResult a = CumulativeParseResult
+    { cumulativeErrors :: [ParseErrorBundle String Void]
+    , cumulativeData :: a
+    }
```

Comment:
This datatype is a bit smelly - the expectation here is that we want to represent a parse that *definitely* succeeded (and thus has a `cumulativeData`), but that also may have some list of errors (which may be empty).

But looking down at `toCumulativeParseResult`, we see `(Monoid a) => ` in the constraints - and `cumulativeData` is set to `mempty`. This means that we have a conflation of states: a parser that succeeded and returns `mempty`, and a parse that fails and we provide `mempty`.

Reply:
OK, as stated elsewhere, this should be more like:

```haskell
type CumulativeParseResult = Either (NonEmpty (ParseErrorBundle String Void))
```

Reply:
... Which is `type ParseResult a` already!